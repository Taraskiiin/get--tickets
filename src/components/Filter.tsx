import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Nav, StyledButton } from '../styles/Filter';
import sortActionCreators from '../duck/TimeOrPriseSort/action-creators';

const Filter = () => {
  const dispatch = useDispatch();

  const sortByPrice = useSelector(
    (store: {
      timeOrPriceReducer: { sortByPrice: boolean; sortByTime: boolean };
    }) => store.timeOrPriceReducer.sortByPrice
  );

  const sortByTime = useSelector(
    (store: {
      timeOrPriceReducer: { sortByPrice: boolean; sortByTime: boolean };
    }) => store.timeOrPriceReducer.sortByTime
  );

  return (
    <Nav className="me-2">
      <StyledButton
        onClick={() => dispatch(sortActionCreators.sortByPriceCreator())}
        active={sortByPrice}
      >
        Самый дешевый
      </StyledButton>
      <StyledButton
        onClick={() => dispatch(sortActionCreators.sortByTimeCreator())}
        active={sortByTime}
      >
        Самый быстрый
      </StyledButton>
    </Nav>
  );
};

export default Filter;
