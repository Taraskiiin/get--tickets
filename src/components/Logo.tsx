import React from 'react';

import { NavLink } from '../styles/Logo';

import logo from '../assets/Logo.png';

const Logo = () => {
  return (
    <NavLink href="#">
      <img src={`${logo}`} alt="logo" />
    </NavLink>
  );
};

export default Logo;
