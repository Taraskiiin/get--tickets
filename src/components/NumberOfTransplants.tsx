import React from 'react';
import { Formik, Field } from 'formik';

import {
  NumberOfTransplantsBlock,
  TitleNumberOfTransplantsBlock,
  FormStyled,
} from '../styles/NumberOfTransplants';

import { useDispatch, useSelector } from 'react-redux';
import { stopSortUpdateCreator } from '../redux/duck/StopsSort';

export interface FormValues {
  selectAll: boolean;
  choosedOption: string[];
}

const NumberOfTransplants: React.FC = () => {
  const dispatch = useDispatch();
  const stops = useSelector(
    (store: {
      stopsSortReducer: {
        stops: FormValues;
      };
    }) => store.stopsSortReducer.stops
  );
  return (
    <NumberOfTransplantsBlock>
      <TitleNumberOfTransplantsBlock>
        Количество пересадок
      </TitleNumberOfTransplantsBlock>
      <Formik
        initialValues={{
          selectAll: stops.selectAll,
          choosedOption: stops.choosedOption,
        }}
        onSubmit={(values: FormValues) => {
          dispatch(stopSortUpdateCreator(values));
        }}
      >
        <FormStyled>
          <label>
            <Field type="checkbox" name="selectAll" />
            &nbsp;Все
          </label>
          <br />
          <label>
            <Field type="checkbox" name="choosedOption" value="0" />
            &nbsp;Без&nbsp;пересадок
          </label>
          <br />
          <label>
            <Field type="checkbox" name="choosedOption" value="1" />
            &nbsp;1&nbsp;пересадка
          </label>
          <br />
          <label>
            <Field type="checkbox" name="choosedOption" value="2" />
            &nbsp;2&nbsp;пересадки
          </label>
          <br />
          <label>
            <Field type="checkbox" name="choosedOption" value="3" />
            &nbsp;3&nbsp;пересадки
          </label>
          <br />
          <button type="submit">Submit</button>
        </FormStyled>
      </Formik>
    </NumberOfTransplantsBlock>
  );
};

export default NumberOfTransplants;
