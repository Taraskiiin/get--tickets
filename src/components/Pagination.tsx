import React, { memo } from 'react';
import { PaginationList, PaginationItem, Link } from '../styles/Pagination';

interface IPaginationProps {
  ticketsPerPage: number;
  totalTickets: number;
  queryParamsState: {
    querySortBy: string | null;
    queryPage: number;
    queryStops: string | null;
  };
}
const Pagination: React.FC<IPaginationProps> = ({
  ticketsPerPage,
  totalTickets,
  queryParamsState,
}) => {
  const pageNumbers = [];
  for (let i = 1; i <= Math.ceil(totalTickets / ticketsPerPage); i++) {
    pageNumbers.push(i);
  }
  return (
    <PaginationList>
      {pageNumbers.map((number) => (
        <Link
          key={number}
          to={`?sortBy=${queryParamsState.querySortBy}&page=${number}&stops=${queryParamsState.queryStops}`}
        >
          <PaginationItem active={queryParamsState.queryPage === number}>
            {number}
          </PaginationItem>
        </Link>
      ))}
    </PaginationList>
  );
};

export default memo(Pagination);
