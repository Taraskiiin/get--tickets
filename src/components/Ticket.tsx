import React, { Fragment } from 'react';

import {
  Card,
  Header,
  Price,
  ColumnInfo,
  ColumnInfoTitle,
  ColumnInfoTime,
  Content,
} from '../styles/Ticket';

export interface ITicket {
  ticket: {
    price: number;
    logo: string;
    rows: {
      from: string;
      to: string;
      arrivesAt: string;
      departsAt: string;
    }[];
  };
}

const Ticket: React.FC<ITicket> = ({ ticket }) => {
  return (
    <Card>
      <Header>
        <Price>{ticket.price}&nbsp;P</Price>
        <img src={ticket.logo} alt="logo" />
      </Header>
      <Content>
        {ticket.rows.map((row, index) => (
          <Fragment key={index}>
            <ColumnInfo>
              <ColumnInfoTitle>
                {row.from} - {row.to}
              </ColumnInfoTitle>
              <span>
                {row.departsAt} - {row.arrivesAt}
              </span>
            </ColumnInfo>
            <ColumnInfo>
              <ColumnInfoTitle>В Пути</ColumnInfoTitle>
              <ColumnInfoTime>
                {/**TODO: add function thst calculates time of the trip */}21ч
                15м
              </ColumnInfoTime>
            </ColumnInfo>
            <ColumnInfo>
              <ColumnInfoTitle>
                Пересадки{/**TODO: add count of transplantation */}{' '}
              </ColumnInfoTitle>
              <span>hkg, jnb</span>
            </ColumnInfo>
          </Fragment>
        ))}
      </Content>
    </Card>
  );
};

export default Ticket;
