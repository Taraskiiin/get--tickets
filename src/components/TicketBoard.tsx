import React, { useEffect } from 'react';

import Ticket, { ITicket } from './Ticket';
import searchIdActionCreators from '../duck/SearchId/action-creators';
import ticketsActionCreators from '../duck/Tickets/action-creators';
import { useDispatch, useSelector } from 'react-redux';

import { useQueryParams } from '../hooks/useQueryParams';

import { sortByPrice } from '../helpers/sortByPrice';
import { sortByTime } from '../helpers/sortByTime';
import { maxValuesStops } from '../helpers/maxValuesStops';

const TicketBoard: React.FC = () => {
  const dispatch = useDispatch();

  const queryParams = useQueryParams();
  const querySortByPrice = queryParams.get('SortByPrice');
  const querySortByTime = queryParams.get('SortByTime');

  const stops = useSelector(
    (store: {
      stopsSortReducer: {
        stops: {
          selectAll: boolean;
          choosedOption: { el: string };
        };
      };
    }) => store.stopsSortReducer.stops
  );

  const searchId = useSelector(
    (store: { searchIdReducer: { searchId: string } }) =>
      store.searchIdReducer.searchId
  );

  const tickets = useSelector(
    (store: { ticketsReducer: { tickets: ITicket[]; stop: boolean } }) =>
      store.ticketsReducer.tickets
  );

  const stop = useSelector(
    (store: { ticketsReducer: { tickets: ITicket[]; stop: boolean } }) =>
      store.ticketsReducer.stop
  );

  useEffect(() => {
    if (stop !== true) {
      searchId === ''
        ? dispatch(searchIdActionCreators.fetchSearchIdCreator())
        : dispatch(ticketsActionCreators.fetchTicketsCreator());
    }
  }, [dispatch, searchId, stop]);

  const filterByStops = tickets?.filter(
    (el) =>
      el.segments[0].stops.length <= maxValuesStops(stops.choosedOption) &&
      el.segments[1].stops.length <= maxValuesStops(stops.choosedOption)
  );

  if (querySortByPrice) {
    sortByPrice(filterByStops);
  } else if (querySortByTime) {
    sortByTime(filterByStops);
  }

  return (
    <div>
      {filterByStops?.length
        ? filterByStops.map((ticket, index) => (
            <Ticket key={index} {...ticket} />
          ))
        : null}
    </div>
  );
};

export default TicketBoard;
