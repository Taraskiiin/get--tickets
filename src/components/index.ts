import Filter from './Filter';
import Logo from './Logo';
import NumberOfTransplants from './NumberOfTransplants';
import Ticket from './Ticket';
import TicketBoard from './TicketBoard';

export { Filter, Logo, NumberOfTransplants, Ticket, TicketBoard };
