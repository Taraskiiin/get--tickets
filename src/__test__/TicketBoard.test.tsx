import React from 'react';
import { TicketBoard } from '../components';
import renderWith from './RenderWith';

describe('<TicketBoard /> testing', () => {
  it('Should render with Redux and Router and ThemeProvider', () => {
    expect(
      renderWith.withRouterAndProviderAndThemeProvider(<TicketBoard />)
    ).toMatchSnapshot();
  });
});
