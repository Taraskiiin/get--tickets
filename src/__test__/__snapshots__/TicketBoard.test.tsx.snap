// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicketBoard /> testing Should render with Redux and Router and ThemeProvider 1`] = `<div />`;
